File: web/src/devReceipts.ts
/** HMAC-SHA256 mint matching server storefront sandbox secrets. */

export const STOREFRONT_DEV = "dev";
export const STOREFRONT_GOOGLE_PLAY = "google_play";
export const STOREFRONT_APP_STORE = "app_store";
export const DEV_STOREFRONT = STOREFRONT_DEV;

export const KNOWN_STOREFRONTS = [
  STOREFRONT_DEV,
  STOREFRONT_GOOGLE_PLAY,
  STOREFRONT_APP_STORE,
] as const;

const SECRET_DEV = "dev-storefront-insecure-secret-change-me";
const SECRET_GOOGLE = "google-play-sandbox-insecure-secret";
const SECRET_APPLE = "app-store-sandbox-insecure-secret";
export const DEFAULT_PACKAGE_NAME = "com.xai.dungeonmaster";

function b64url(bytes: ArrayBuffer | Uint8Array): string {
  const u8 = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  let s = "";
  for (let i = 0; i < u8.length; i++) s += String.fromCharCode(u8[i]);
  return btoa(s).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

export function secretFor(storefront: string): string {
  switch (storefront.toLowerCase()) {
    case STOREFRONT_GOOGLE_PLAY:
      return SECRET_GOOGLE;
    case STOREFRONT_APP_STORE:
      return SECRET_APPLE;
    default:
      return SECRET_DEV;
  }
}

export async function signDevReceipt(
  productId: string,
  secret = SECRET_DEV,
): Promise<string> {
  const enc = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    enc.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const product = enc.encode(productId);
  const sig = await crypto.subtle.sign("HMAC", key, product);
  return `${b64url(product)}.${b64url(sig)}`;
}

export type MintedReceipt = {
  storefront: string;
  productId: string;
  receipt: string;
};

/** Mint a verify-ready receipt for the given storefront (JSON for Play / App Store). */
export async function mintReceipt(
  storefront: string,
  productId: string,
  packageName = DEFAULT_PACKAGE_NAME,
): Promise<MintedReceipt> {
  const id = (storefront || STOREFRONT_DEV).toLowerCase();
  const hmac = await signDevReceipt(productId, secretFor(id));
  let receipt = hmac;
  if (id === STOREFRONT_GOOGLE_PLAY) {
    receipt = JSON.stringify({
      packageName,
      productId,
      purchaseToken: hmac,
    });
  } else if (id === STOREFRONT_APP_STORE) {
    receipt = JSON.stringify({
      receiptData: hmac,
      productId,
    });
  }
  return { storefront: id, productId, receipt };
}

File: android/app/src/main/java/com/xai/dungeonmaster/android/DevReceipts.kt
package com.xai.dungeonmaster.android

import android.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Client-side mint of sandbox receipts for server storefront plugins.
 *
 * HMAC format (shared by `dev`, and sandbox modes of `google_play` / `app_store`):
 * `base64url(productId).base64url(HMAC_SHA256(secret, productId))`
 *
 * Defaults match the engine's insecure local secrets so the Store tab works
 * out of the box. For Google Play / App Store, the payload is wrapped as the
 * JSON shape live clients will eventually send (purchaseToken / receiptData).
 */
object DevReceipts {

    const val STOREFRONT_DEV = "dev"
    const val STOREFRONT_GOOGLE_PLAY = "google_play"
    const val STOREFRONT_APP_STORE = "app_store"

    /** @deprecated use [STOREFRONT_DEV] */
    const val STOREFRONT_ID = STOREFRONT_DEV

    const val SECRET_DEV = "dev-storefront-insecure-secret-change-me"
    const val SECRET_GOOGLE = "google-play-sandbox-insecure-secret"
    const val SECRET_APPLE = "app-store-sandbox-insecure-secret"

    /** Default package / bundle used in JSON wrappers (override in real builds). */
    const val DEFAULT_PACKAGE_NAME = "com.xai.dungeonmaster"

    /** @deprecated use [SECRET_DEV] */
    const val DEFAULT_SECRET = SECRET_DEV

    val knownStorefronts: List<String> = listOf(
        STOREFRONT_DEV,
        STOREFRONT_GOOGLE_PLAY,
        STOREFRONT_APP_STORE,
    )

    fun secretFor(storefront: String): String = when (storefront.lowercase()) {
        STOREFRONT_GOOGLE_PLAY -> SECRET_GOOGLE
        STOREFRONT_APP_STORE -> SECRET_APPLE
        else -> SECRET_DEV
    }

    fun sign(productId: String, secret: String = SECRET_DEV): String {
        val p = productId.toByteArray(Charsets.UTF_8)
        val body = b64(p)
        val sig = b64(hmac(secret.toByteArray(Charsets.UTF_8), p))
        return "$body.$sig"
    }

    /**
     * Build a receipt + storefront id pair ready for `POST /v2/entitlements/verify`.
     * Google Play and App Store use JSON envelopes so the payload matches live clients.
     */
    fun mint(storefront: String, productId: String, packageName: String = DEFAULT_PACKAGE_NAME): Minted {
        val id = storefront.ifBlank { STOREFRONT_DEV }.lowercase()
        val hmac = sign(productId, secretFor(id))
        val receipt = when (id) {
            STOREFRONT_GOOGLE_PLAY ->
                """{"packageName":${json(packageName)},"productId":${json(productId)},"purchaseToken":${json(hmac)}}"""
            STOREFRONT_APP_STORE ->
                """{"receiptData":${json(hmac)},"productId":${json(productId)}}"""
            else -> hmac
        }
        return Minted(storefront = id, productId = productId, receipt = receipt)
    }

    data class Minted(val storefront: String, val productId: String, val receipt: String)

    private fun json(s: String): String =
        "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\""

    private fun hmac(secret: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret, "HmacSHA256"))
        return mac.doFinal(data)
    }

    private fun b64(bytes: ByteArray): String =
        Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
}
